python/lsst/daf/butler/_quantum_backed.py

```diff
@@ -399,26 +406,112 @@ class QuantumProvenanceData(pydantic.BaseModel):
     when the QuantumGraph was built.
     """
 
-    actual_outputs: Set[uuid.UUID]
+    actual_outputs: Set[DatasetId]
     """Unique IDs of datasets that were actually written when this quantum
     was executed.
     """
 
-    locations: Dict[str, Set[uuid.UUID]] = defaultdict(set)
+    locations: Dict[str, Set[DatasetId]]
     """Mapping from datastore name to the set of `actual_output` dataset IDs
     written by this quantum.
     """
 
-    records: Dict[str, List[StoredDatastoreItemInfo]] = defaultdict(list)
+    records: Dict[str, List[StoredDatastoreItemInfo]]
     """Rows from the opaque tables used by datastores for the `actual_output`
     datasets written by this quantum, indexed by opaque table name.
     """
 
+    def to_simple(self, minimal: bool = False) -> Dict[str, Any]:
+        """Make representation of the provenance suitable for serialization."""
+        # dataclasses.asdict does not know how to handle some types, have to
+        # do it manually. Also have to replace sets with lists as some
+        # serializers do not support set type.
+        def _serialize_dataset_id(id: DatasetId) -> Union[int, str]:
+            return id if isinstance(id, int) else f"urn:uuid:{id}"
+
+        def _serialize_dataset_ids(ids: Set[DatasetId]) -> List[Union[int, str]]:
+            return [_serialize_dataset_id(id) for id in ids]
+
+        records: Dict[str, List[Dict[str, Any]]] = {}
+        for table_name, table_records in self.records.items():
+            records[table_name] = []
+            for record in table_records:
+                record_dict = record.to_record()
+                # Have to remember actual class name of the record.
+                record_dict["__class__"] = get_full_type_name(record)
+                if "dataset_id" in record_dict:
+                    record_dict["dataset_id"] = _serialize_dataset_id(record_dict["dataset_id"])
+                records[table_name].append(record_dict)
+        locations = {datastore: _serialize_dataset_ids(ids) for datastore, ids in self.locations.items()}
+        return dict(
+            predicted_inputs=_serialize_dataset_ids(self.predicted_inputs),
+            available_inputs=_serialize_dataset_ids(self.available_inputs),
+            actual_inputs=_serialize_dataset_ids(self.actual_inputs),
+            predicted_outputs=_serialize_dataset_ids(self.predicted_outputs),
+            actual_outputs=_serialize_dataset_ids(self.actual_outputs),
+            locations=locations,
+            records=records,
+        )
+
+    @classmethod
+    def from_simple(
+        cls,
+        simple: Dict[str, Any],
+        universe: Optional[DimensionUniverse] = None,
+        registry: Optional[Registry] = None,
+    ) -> QuantumProvenanceData:
+        """Make an instance of this class from serialized data.
+
+        Parameters
+        ----------
+        data : `dict`
+            Serialized representation returned from `to_simple` method,
```

Comment:
Missing some params, but I also don't see them used.  Just remove them from the signature?  I know we've got a bit of a convention here for which optional arguments these methods should take, but I don't think they're actually called in any completely generic context where the caller can't know the exact signature.  I suppose the same applies to `minimal` in `to_simple`.  @timj, do you have opinions on whether to have these arguments when they aren't used?

Reply:
We do have a convention that these to_simple flags can propagate down to places that might use them. This is because we wanted to be able to say "give me minimal stuff" and know that it will either be ignored or include just the minimum. I think this means we need to keep the unused parameters around and be consistent in all to_simple implementations.